cli/src/commands/workspace.rs

```diff
@@ -280,20 +281,73 @@ fn cmd_workspace_root(
     Ok(())
 }
 
+fn create_and_check_out_recovery_commit(
+    command: &CommandHelper,
+    ui: &mut Ui,
+    workspace: Workspace,
+) -> Result<Arc<ReadonlyRepo>, CommandError> {
+    let operation = command.resolve_operation(ui, workspace.repo_loader())?;
+    let repo = workspace.repo_loader().load_at(&operation)?;
+    let workspace_id = workspace.workspace_id().clone();
+    let tree_id = repo.store().empty_merged_tree_id();
+
+    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo.clone())?;
+    let (mut locked_workspace, commit) =
+        workspace_command.unchecked_start_working_copy_mutation()?;
+    let commit_id = commit.id();
+
+    let mut tx = repo.start_transaction(command.settings());
+    let mut_repo = tx.mut_repo();
+    let new_commit = mut_repo
+        .new_commit(command.settings(), vec![commit_id.clone()], tree_id.clone())
+        .write()?;
+    mut_repo.set_wc_commit(workspace_id, new_commit.id().clone())?;
+    let ret = Ok(tx.commit("recovery commit"));
+
+    locked_workspace.locked_wc().reset_to_empty()?;
+    locked_workspace.finish(operation.id().clone())?;
+
+    writeln!(
+        ui.stderr(),
+        "Created and checked out recovery commit {}",
+        short_commit_hash(new_commit.id())
+    )?;
+
+    ret
+}
+
 /// Loads workspace that will diverge from the last working-copy operation.
 fn for_stale_working_copy(
     command: &CommandHelper,
     ui: &mut Ui,
-) -> Result<WorkspaceCommandHelper, CommandError> {
+) -> Result<(WorkspaceCommandHelper, bool), CommandError> {
     let workspace = command.load_workspace()?;
     let op_store = workspace.repo_loader().op_store();
-    let op_id = workspace.working_copy().operation_id();
-    let op_data = op_store
-        .read_operation(op_id)
-        .map_err(|e| CommandError::InternalError(format!("Failed to read operation: {e}")))?;
-    let operation = Operation::new(op_store.clone(), op_id.clone(), op_data);
-    let repo = workspace.repo_loader().load_at(&operation)?;
-    command.for_loaded_repo(ui, workspace, repo)
+    let (repo, recovered) = {
+        let op_id = workspace.working_copy().operation_id();
+        match op_store.read_operation(op_id) {
+            Ok(op_data) => (
+                workspace.repo_loader().load_at(&Operation::new(
+                    op_store.clone(),
+                    op_id.clone(),
+                    op_data,
+                ))?,
+                false,
+            ),
+            Err(e) => {
```

Comment:
We'll also need to test if `e` is of `ObjectNotFound` type.

But the user might want to continue the recovery in that situation. That's another reason why I think an explicit command option is better.

Reply:
I've added the test.

What do you mean by "continue the recovery"? The user has already initiated the recovery (by running "workspace update-stale") so presumably we don't need to ask the user whether they want to continue, since their intention is probably to do the recovery to the end.

Reply:
The user might want to try `update-stale` on some other kind of errors. If the error is `ObjectNotFound`, it's probably safe to continue. If the error isn't, `update-stale` might not be always the right way for recovery, but he can try.

Reply:
Ah, I see. So I think that adding the test (which I have done) gets what we want (if it's `ObjectNotFound`, it's safe and we can continue, but if not, we tell the user).

Reply:
> it's `ObjectNotFound`, it's safe and we can continue, but if not, we tell the user).

And the user might still want to try `update-stale` in that situation. That's what I wanted to say.